feat: support cluster.x-k8s.io/paused annotation

devigned opened the pull request:
**What this PR does / why we need it**:

As specified in cluster API alpha2 to alpha3, providers need to check for pause annotation in cluster/machine controller which is used by clusterctl during a move. This is not present in azure provider. This PR add an early return from each controller if the cluster is paused. It also adds a watch for cluster unpause in both AzureManagedCluster and AzureCluster.

**Testing:**
As part of testing this PR, I wanted to be able to see what was going on in the controller when reconcile doesn't actually change the state of the resource. One way of doing this which is included in this PR is by inspecting log entries for artifacts which prove the execution path based on the test setup. I think this also pays dividends to prove that an operator of the controller would be able to observe what is happening in the controller at a given log level.

**TODO:**
- [x] add tests to AzureCluster and AzureMachine reconciler
- [x] add tests to AzureManagedCluster, AzureMachinePool and AzureManagedMachinePool

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #601

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support cluster.x-k8s.io/paused annotation which is used during cluster move
```

devigned commented:
/test pull-cluster-api-provider-azure-e2e

devigned commented:
/retest

devigned commented on controllers/azuremachine_controller_test.go:
```diff
@@ -48,5 +48,9 @@ var _ = Describe("AzureMachineReconciler", func() {
 			Expect(result.RequeueAfter).To(BeZero())
 		})
+
+		It("should exit early if the cluster is paused", func() {
```
@nader-ziada, I'd love to get your perspective on this test now that it's filled in.

devigned commented:
@CecileRobertMichon, if you have a sec, I'd appreciate your perspective on the testing add thus far in this PR. I am going to hold off on implementing the same thing with the exp controllers.

CecileRobertMichon commented on controllers/azurecluster_controller_test.go:
```diff
@@ -87,5 +94,66 @@ var _ = Describe("AzureClusterReconciler", func() {
 			Expect(err.Error()).To(Or(Equal("context deadline exceeded"), Equal("rate: Wait(n=1) would exceed context deadline")))
 		})
+
+		It("should fail skip reconciliation if cluster is paused", func() {
```
```suggestion
		It("should skip reconciliation if cluster is paused", func() {
```
?

devigned replied:
fixed

CecileRobertMichon commented:
@devigned I like this approach. Is there an easy way to test for log lines with different log levels?

CecileRobertMichon commented:
@vincepri @detiber if you have time for a quick look, any feedback on this controller testing approach?

devigned commented:
Great question, @CecileRobertMichon! `LogEntry.Level` should correspond to the log level of the event (e.g. `log.V(1)` should be `LogEntry.Level == 1`).

Looks like that might not be working... I'll look into it.

detiber commented on controllers/azurecluster_controller_test.go:
```diff
@@ -87,5 +94,66 @@ var _ = Describe("AzureClusterReconciler", func() {
+				Expect(err).NotTo(HaveOccurred())
+			}()
+
+			Eventually(logListener.GetEntries).Should(ContainElement(
```
Might be good to avoid trying to match on specific log entries in favor of trying to determine by the resource itself to see if things were short-circuited

devigned replied:
100% agree. Until we have ObservedVersion on the resources, this seems like the only way to verify. Am I missing a better way?

nader-ziada replied:
I also can't think of other validations at the moment, but I worry that we are adding a lot of code to enable checking of the logs in tests  when its not the most reliable

devigned replied:
Hmm... Why do you say that logs are not reliable? The names of the resources are unique and part of the criteria for the test.

Also, I kind of like how verifying a log output is like taking the role of a cluster operator viewing logs looking for an indication of the thing happening. It's very similar to how I would start out debugging a problem in the running controller.

nader-ziada replied:
I'm saying that because any change can make the log message different, for example a different error being returned or even a slight change in the message of a lower level function

nader-ziada replied:
I've had to use logs output before as well, sometimes there is no other way

devigned commented:
Okay, I've added `LogEntry.LogFunc` which would be equal to "Info" or "Error" depending on which one was called. if `log.V(2).Info(...)` is called, `LogEntry` would be the following:
```go
LogEntry{
    Level: 2,
    LogFunc: "Info"
    Values: ...
}
```

detiber commented on exp/controllers/azuremangedcontrolplane_controller.go:
```diff
@@ -24,6 +24,8 @@ import (
+	"sigs.k8s.io/cluster-api/util/annotations"
+
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
 	infrav1exp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
```
SetupWithManager should be updated to:
- Add `WithEventFilter(predicates.ResourceNotPaused(r.Log)).` to the `ctrl.NewControllerManagedBy()` chain to filter out reconciliations when the paused annotation is set on any of the resources being watched.
- Add a Watch for `Cluster` that uses `predicates.ClusterUnpausedAndInfrastructureReady()` to properly trigger reconciliation when Cluster infrastructure becomes ready and when the Cluster is unpaused, similar to what is being done here: https://github.com/kubernetes-sigs/cluster-api/blob/2bb56ed2eb2a91cb59b740cb286992952da3dfed/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L121-L130

devigned commented on controllers/azurecluster_controller.go:
```diff
@@ -46,10 +51,27 @@ type AzureClusterReconciler struct {
 		WithOptions(options).
 		For(&infrav1.AzureCluster{}).
-		Complete(r)
+		Build(r)
```
Fixed.

devigned commented on controllers/azuremachine_controller.go:
```diff
@@ -34,6 +34,7 @@ import (
+	"sigs.k8s.io/cluster-api/util/annotations"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
```
Fixed.

devigned commented:
/hold cancel

devigned commented:
API Diff:
```
sigs.k8s.io/cluster-api-provider-azure/exp/controllers
  Compatible changes:
  - AzureClusterToAzureMachinePoolsMapper: added
  - AzureManagedClusterToAzureManagedControlPlaneMapper: added
  - AzureManagedClusterToAzureManagedMachinePoolsMapper: added
  - MachinePoolToInfrastructureMapFunc: added

sigs.k8s.io/cluster-api-provider-azure/controllers
  Incompatible changes:
  - (*AzureMachineReconciler).AzureClusterToAzureMachines: removed
  Compatible changes:
  - AzureClusterToAzureMachinesMapper: added
  - GetObjectsToRequestsByNamespaceAndClusterName: added
  - GetOwnerClusterName: added
```

devigned commented:
I believe I have addressed feedback. Thank you for the guidance, @detiber and @nader-ziada.

nader-ziada commented:
lgtm, 
@detiber can you please take a final look

detiber commented on controllers/azurecluster_controller.go:
```diff
@@ -46,10 +51,28 @@ type AzureClusterReconciler struct {
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: util.ClusterToInfrastructureMapFunc(azCluster.GroupVersionKind()),
+		},
+		predicates.ClusterUnpausedAndInfrastructureReady(r.Log),
```
This should just check for unpaused, otherwise we might filter out events that we want to act on, since this controller is responsible for providing the value that populates cluster.status.infrastructureReady
```suggestion
		predicates.ClusterUnpaused(r.Log),
```

devigned replied:
Changed to unpaused only. TY

detiber commented on exp/controllers/azuremanagedcluster_controller.go:
```diff
@@ -45,10 +51,27 @@ type AzureManagedClusterReconciler struct {
+		return errors.Wrapf(err, "error creating controller")
+	}
+
+	// Add a watch on clusterv1.Cluster object for unpause & ready notifications.
```
This should just check for unpaused, otherwise we might filter out events that we want to act on, since this controller is responsible for providing the value that populates cluster.status.infrastructureReady
```suggestion
	// Add a watch on clusterv1.Cluster object for unpause notifications.
```

devigned replied:
Changed to unpaused only.

detiber commented on exp/controllers/azuremanagedcluster_controller.go:
```diff
@@ -45,10 +51,27 @@ type AzureManagedClusterReconciler struct {
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: util.ClusterToInfrastructureMapFunc(azManagedCluster.GroupVersionKind()),
+		},
+		predicates.ClusterUnpausedAndInfrastructureReady(r.Log),
```
This should just check for unpaused, otherwise we might filter out events that we want to act on, since this controller is responsible for providing the value that populates cluster.status.infrastructureReady
```suggestion
		predicates.ClusterUnpaused(r.Log),
```

devigned replied:
Changed to unpaused only.

detiber commented on exp/controllers/azuremangedcontrolplane_controller.go:
```diff
@@ -24,17 +24,24 @@ import (
 	"github.com/pkg/errors"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/client-go/tools/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
 	infrav1exp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
 	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 
-	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
 	"sigs.k8s.io/cluster-api/util"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
```
nit: The imports here should be re-grouped and re-sorted

devigned replied:
Fixed

detiber commented on exp/controllers/azuremanagedcluster_controller.go:
```diff
@@ -25,6 +25,12 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
 	infrav1exp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
 	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 
```
nit: imports should likely be re-grouped and re-sorted

devigned replied:
Fixed

detiber commented on exp/controllers/azuremachinepool_controller.go:
```diff
@@ -43,7 +45,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
-	capzcntr "sigs.k8s.io/cluster-api-provider-azure/controllers"
+	controllers "sigs.k8s.io/cluster-api-provider-azure/controllers"
 	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
```
nit: imports should likely be re-grouped and re-sorted

devigned replied:
Resorted. TY

detiber commented on controllers/azuremachine_controller.go:
```diff
@@ -34,6 +35,7 @@ import (
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
 	capierrors "sigs.k8s.io/cluster-api/errors"
 	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
```
nit: imports should likely be re-grouped and re-sorted

devigned replied:
Resorted. TY

detiber commented on controllers/azurecluster_controller.go:
```diff
@@ -24,17 +24,22 @@ import (
 	"github.com/pkg/errors"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/client-go/tools/record"
+
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
 	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
 	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
 	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/predicates"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
```
nit: imports should likely be re-grouped and re-sorted

devigned replied:
Fixed. Oddly, wasn't showing up in my dev environment.

devigned commented on controllers/azurecluster_controller.go:
```diff
@@ -46,10 +51,28 @@ type AzureClusterReconciler struct {
+		return errors.Wrapf(err, "error creating controller")
+	}
+
+	// Add a watch on clusterv1.Cluster object for unpause & ready notifications.
```
Great catch. TY!

detiber commented:
/lgtm
